### PR TITLE
Restore removed API from Extensions

### LINKED ISSUE #87480: [API Proposal]: Bring back extensions API to ease adoption of extensions libraries from older frameworks
### Background and motivation

Over time types have been removed from the Microsoft.Extensions.* libraries that have made it difficult for customers to use newer versions of those packages on older frameworks.

All of these APIs were obsoleted in a predictable way - first marking as obsolete for one release, then removal in a following release - however that removal has proven to be more breaking for the ecosystem in some cases.

The largest problem that has come up a lot is the removal of types that were used by ASP.NET Core in 2.x.  Many of the libraries from this release are still the latest in use by .NETFramework customers - since that's the last version of the product that supported .NETFramework.

Bringing back these APIs has a small cost to maintenance, but will unblock folks from taking newer extensions packages and will allow other packages built on top of these extensions to safely take on the latest version without regressing people.  It will also simplify our story for the Microsoft.Extesions packages in terms of best practice reccomendations.  We can tell folks it's safe to reference the latest version that is supported and has the API they wish to use (same as we recommend for `System.Text.Json`) and not require folks to specialize and cross-target for each TFM to select a different version.

We've chosen to bring back API as 
```c#
[EditorBrowsable(EditorBrowsableState.Never)]
[Obsolete("This property is retained only for compatibility. ...", error: true)]
```
To discourage use.  This will fix binary compatibility, reduce discoverability, source compatibility is still broken by default but has the workaround of annotating consuming code with obsolete.

### API Proposal

See https://github.com/dotnet/runtime/pull/85846/files?file-filters%5B%5D=.cs&show-viewed-files=true for added API

API added back is not subject to review since it must be identical to what was previously exposed.

I'm seeking approval / acknowledgement for adding back this API for the sake of binary compatibility.

We looked at all the API breaking changes between 2.1 and today and choose to add back API based on the following:
1. Entry point API - extension methods, public properties on types we expect users to interact with.
2. High usage

API was not brought back based on the following:
1. Bringing back would break a runtime scenario like DI
2. Pub-ternals which we never expect users to interact with
3. Concrete types where we expect users to interact with interfaces

We double checked any case where we were excluding API and it's usage showed significant numbers to understand if there was a significant use case we were missing.  One case of this was `NullScope` where we found all the usage was from tool-like packages redistributing the Microsoft.Extensions binaries themselves as an app layout - and is this not subject to break through package upgrade.

Usage analysis: [xls](https://microsoft.sharepoint.com/:x:/t/netfx/corefx/EfNGUsnG6BFJmye5nJvu7oUBvdyzdT_Z_1lUzsCaGzetcw?e=gFTdm8) [md](https://gist.github.com/ericstj/c06988661190ec62736c748316692675)

### API Usage

N/A

### Alternative Designs

Only do this in implementation assemblies and not reference assemblies.  This isn't helpful since we don't ship the reference assemblies in nuget packages.

### Risks

We may not have brought back enough API - we can always bring back more based on customer feedback.  We have tested some scenarios (like 2.1 ASP.NET Core App on NetFx with the latest packages - fixed).

Normal set of risk with adding API - name collisions, reflection ambiguity, etc.  We've tried to anticipate some of this - like the DI consumption of constructors and reduced the API exposed to mitigate risk.

### PR DESCRIPTION
**Pull request by ericstj**
Fixes https://github.com/dotnet/runtime/issues/87480
I spent some time seeing what it would look like to bring back API that was present on extensions packages from 2.1-3.1 so that libraries built against these old packages would at least bind against the latest packages.

I stubbed out most of the API that was added back, unless it seemed "easy" to make work or important.  

Have a look below and comment about the cases and if you think we should do more (or less).

This covers most of the cases.  There's one case I cannot fix:
```
CP0006: Cannot add interface member 'Microsoft.Extensions.Hosting.IHostBuilder.UseServiceProviderFactory<TContainerBuilder>(System.Func<Microsoft.Extensions.Hosting.HostBuilderContext, Microsoft.Extensions.DependencyInjection.IServiceProviderFactory<TContainerBuilder>>)' to e7\bin\Debug\net4.8\Microsoft.Extensions.Hosting.Abstractions.dll because it does not exist on e2\bin\Debug\net4.8\Microsoft.Extensions.Hosting.Abstractions.dll
```
This is an added method to an interface.  One way to _fix_ this would be to go back and modify any type that implemented that interface to add the method as "public virtual" (luckily the signature doesn't require any new types).  I was able to test this and it is sufficient for the runtime to accept that method as satisfying the interface -- even if the implementer didn't see that method as part of the interface at compile time.  Chances are [there are very few implementations](https://grep.app/search?q=class%20.%2AIHostBuilder&regexp=true) of `IHostBuilder` out there - so it might be tractable to patch them (or even just ignore them).

### CODE REVIEW

**Comment by joperezr** [comment]
FYI @geeknoid

File: src/libraries/Microsoft.Extensions.Caching.Memory/ref/Microsoft.Extensions.Caching.Memory.cs
Comment (joperezr) [inline src/libraries/Microsoft.Extensions.Caching.Memory/ref/Microsoft.Extensions.Caching.Memory.cs:45]: Is this just the default obsolete message or do we actually mean the fact that we want to remove in the future?

File: src/libraries/Microsoft.Extensions.Hosting/ref/Microsoft.Extensions.Hosting.cs
Comment (joperezr) [inline src/libraries/Microsoft.Extensions.Hosting/ref/Microsoft.Extensions.Hosting.cs:131]: NIT: Should this have an obsolete suggesting to use the other two ctors?

File: src/libraries/Microsoft.Extensions.Logging.Console/ref/Microsoft.Extensions.Logging.Console.cs
Comment (joperezr) [inline src/libraries/Microsoft.Extensions.Logging.Console/ref/Microsoft.Extensions.Logging.Console.cs:32]: NIT: I'm assuming the removal of the trimmer attributes is unintentional?

**Comment by tarekgh** [comment]
Is it possible to have all re-added obsolete APIs be in one source file inside each component instead of scattering it across the code? using partial classes can do it I guess.

File: src/libraries/Microsoft.Extensions.Caching.Memory/ref/Microsoft.Extensions.Caching.Memory.cs
Comment (ericstj) [reply src/libraries/Microsoft.Extensions.Caching.Memory/ref/Microsoft.Extensions.Caching.Memory.cs:45]: This was the message that was on these members before they were removed.

**Comment by ericstj** [comment]
> Is it possible to have all re-added obsolete APIs be in one source file inside each component instead of scattering it across the code? using partial classes can do it I guess.

I did demonstrate this method in some of the assemblies, but I'm not sold on that being preferable.  Also if you noticed - most of the stuff we brought back was entire new types, very few members removed or renamed.

Rather than look at the content / actual mechanism I obsoleted code, have a look at the entire mess of types we're bringing back and weigh in on which we think would actually be meaningful to bring back.

IMO all the logging stuff looks pretty busted.  There is a lot we had to bring back and none of it "works" -- except for the internal types I had to make public.  I have a hard time believing that folks which depended on this would be unblocked by adding back throwing API.  Perhaps I could 1) invest more time in understanding if we could make it functional, 2) make it noop instead of throw, 3) cut bringing back all the logging.

Everything but logging looks pretty "easy".  We can bring it back in a way that seems likely to unblock folks that used it.

File: src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleLogger.cs
Comment (ericstj) [inline src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleLogger.cs:18]: These types have a ton of breaking changes, so it might be better to just expose empty types and rename these internal ones, or not bring back any of logging.

File: src/libraries/Microsoft.Extensions.Caching.Memory/src/MemoryCacheOptions.cs
Comment (geeknoid) [inline src/libraries/Microsoft.Extensions.Caching.Memory/src/MemoryCacheOptions.cs:45]: In general, is this the right message? Maybe instead:

"Present for backward compatibility only, please do not use in new code"

?

File: src/libraries/Microsoft.Extensions.Configuration/src/ConfigurationSection.cs
Comment (geeknoid) [inline src/libraries/Microsoft.Extensions.Configuration/src/ConfigurationSection.cs:38]:
```suggestion
        [Obsolete]
        public ConfigurationSection(ConfigurationRoot root, string path) : this((IConfigurationRoot)root, path)
```

File: src/libraries/Microsoft.Extensions.Hosting.Abstractions/src/IApplicationLifetime.cs
Comment (geeknoid) [inline src/libraries/Microsoft.Extensions.Hosting.Abstractions/src/IApplicationLifetime.cs:18]: Isn't deleting all these members a breaking change?

File: src/libraries/Microsoft.Extensions.Caching.Memory/src/MemoryCacheOptions.cs
Comment (ericstj) [reply src/libraries/Microsoft.Extensions.Caching.Memory/src/MemoryCacheOptions.cs:45]: None of these messages have been scrubbed - I just brought back them as they were in the last release or put a TODO in place.  This PR isn't meant to review for style/final merge.  It's sharing an outline to drive discussion on next steps.

File: src/libraries/Microsoft.Extensions.Hosting.Abstractions/src/IApplicationLifetime.cs
Comment (ericstj) [reply src/libraries/Microsoft.Extensions.Hosting.Abstractions/src/IApplicationLifetime.cs:18]: They aren't deleted, they are "pushed down".  What happened in the past was the interface was renamed.  At first I added them back but then realized I couldn't use them in other API.  So I instead added them back as derived from their new name.  We use this trick in classes - pushing a member to a base class is a compatible change (both at compile and runtime).  I tested and this works for implicitly implemented interfaces, but it looks like the runtime isn't happy with it if the interface is explicitly implemented.  Probably that could be fixed, but not in the places we care about.  That makes this approach a no-go.

I could do the same thing with adapters.  Alternatively we could just throw from API that used these interfaces and not try to implement it.

File: src/libraries/Microsoft.Extensions.Configuration/src/ConfigurationSection.cs
Comment (ericstj) [reply src/libraries/Microsoft.Extensions.Configuration/src/ConfigurationSection.cs:38]: We cannot add obsolete here since it would be source-breaking.  The compiler would prefer the concrete type over interface overload.  I don't think there is any harm in having this overload.

https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBLANmgExAGoAfAAQCYBGAWACgGA3AQygAIx2BedgOxgB3dgGEAFAEoA3A2Zt2wHvyHsAQmLDTZjeuQDM7KqPYh2ASQBiLMDAYBvBgF9t+9tj4YYUAGbWY5qxt2O2cdVyNVewZ2GMMDdUs/dl8bCWDQ2PYAbQB5YABnCFwYTwBdaNjXdRFONNMMAAtsfLExRNTNdO1HIA===

File: src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleLogger.cs
Comment (ericstj) [reply src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleLogger.cs:18]: I wonder if the balance we might strike in logging is not to expose any of the concrete loggers or related helper types/interfaces.  Only expose the extension methods.  I might prototype this to see if that's doable.  Still looking for usage information to validate this approach.

**Comment by geeknoid** [comment]
@RussKie has been working to get R9 samples working and encountering compat issues related to the fact old R9 had graduated dependencies (where the R9 bits for net462 depended on MS extensions 2.2, but the R9 bits for net7.0 depended on MS extensions 7.0, etc.), whereas the newer R9 now depends exclusively on MS extensions 8.0.

So these R9 samples could be used as a good basic vetting of whether the compat fixes actually make stuff work....

File: src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleLogger.cs
Comment (ericstj) [reply src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleLogger.cs]: This is the approach I'm taking.  Only bringing back entry-point APIs and not these pub-ternal like types.

**Comment by ericstj** [comment]
> So these R9 samples could be used as a good basic vetting of whether the compat fixes actually make stuff work....

I looked at this failure and its ASP.NETCore 2.x on .NET Framework, I suspect all the blockers will originate from this scenario and less R9's usage of extensions.  I was able to unblock it with the bits from this PR, but not certain about the amount of coverage that sample has.  Also not sure if we want to focus on this scenario for .NET consumption - it might be better to show usage in classic ASP.NET since that's where more customers will be on that framework.

File: src/libraries/Microsoft.Extensions.Hosting/src/Internal/ConsoleLifetime.obsolete.cs
Comment (ericstj) [inline src/libraries/Microsoft.Extensions.Hosting/src/Internal/ConsoleLifetime.obsolete.cs:14]: This addition is causing many tests to fail with: 
```
Void .ctor(Microsoft.Extensions.Options.IOptions`1[Microsoft.Extensions.Hosting.ConsoleLifetimeOptions], Microsoft.Extensions.Hosting.IHostEnvironment, Microsoft.Extensions.Hosting.IHostApplicationLifetime, Microsoft.Extensions.Options.IOptions`1[Microsoft.Extensions.Hosting.HostOptions], Microsoft.Extensions.Logging.ILoggerFactory)
Void .ctor(Microsoft.Extensions.Options.IOptions`1[Microsoft.Extensions.Hosting.ConsoleLifetimeOptions], Microsoft.Extensions.Hosting.IHostingEnvironment, Microsoft.Extensions.Hosting.IApplicationLifetime)
```

Comment (ericstj) [reply src/libraries/Microsoft.Extensions.Hosting/src/Internal/ConsoleLifetime.obsolete.cs:14]: Given this type is internal, and API usage is literally non-existent I'm going to pull this one out:
https://apisof.net/catalog/74559930-0e33-2f5b-82ad-66e031022b84

**Comment by ericstj** [comment]
Some information about using these packages on different versions of .NET.

| Framework | In support? | Works with 8.0 packages? | Works with this PR? | Notes |
| ------------ | ------------ | ----------------------------- | ---------------------- | ------ |
| NETFramework (net462+) | yes | yes | yes | No dependencies on extensions by default | 
| ASP.NETCore on NETFramework (net462+) | yes* | no | yes | While this does enable upgrade of extensions packages here, we  document that we don't support that upgrade https://dotnet.microsoft.com/en-us/platform/support/policy/aspnet/2.1-packages | 
| netcoreapp2.1 | no | no | yes | Avoids problem with 3.x due to not calling WebHost instead of generic host APIs |
| netcoreapp3.1 | no | no | somewhat | These packages allow the app to get further, but it fails to start on Windows due to EventLog dropping support for netcoreapp3.1 | 
| net5.0 | no | somewhat | somewhat | Fails to start on Windows for same reason as 3.1 | 
| net6.0 | yes | yes | yes | |
| net7.0 | yes | yes | yes | |

So really the only place that this PR is making a difference is on unsupported frameworks or for unsupported upgrades.

**Comment by geeknoid** [comment]
But the point of these changes is not to show that net462 code can work with .NET 8 packages, it's about making it possible for an application developed using 2.2 extensions will work when the 8.0 extension are substituted under. If there are breaking changes between 2.2 and 8.0, then this is not possible.

So can a net462 app compiled using the 2.2 version of MS Extensions work with the 8.0 version of MS Extensions?

**Comment by ericstj** [comment]
> So can a net462 app compiled using the 2.2 version of MS Extensions work with the 8.0 version of MS Extensions?

It depends on what they used.  If they didn't use any of the obsoleted API it'd be fine.  We need to see what "typical" usage of extensions looks like on net462 that is not ASP.NET Core on NETFramework (which is the only reason for folks to stay on 2.1 -- in which case we are not interested in updating them per above).  Also - 2.2 packages have been out of support for many years now.  There's really no reason for folks to stay on those versions.  There is no "platform affinity" between NETFramework versions and any specific release of Microsoft.Extensions nuget packages, so we shouldn't be thinking about pulling folks forward on NETFramework - those are just the latest packages with the latest features and fixes - just like any other package.

**Comment by ericstj** [comment]
This is now ready for review.  The final decision on in/out API and justification can be found here:
[MSFT internal xls](https://microsoft.sharepoint.com/:x:/r/teams/netfx/corefx/_layouts/15/Doc.aspx?sourcedoc=%7BC95246F3-E8C6-4911-9B27-B99C9BEEEE85%7D&file=extensions-compat-usage.xlsx&action=default&mobileredirect=true&share=IQHzRlLJxugRSZsnuZyb7u6FAQe-PcNae_gSYRixThXzBmY) / [public gist](https://gist.github.com/ericstj/c06988661190ec62736c748316692675)
You can also see the differences with what we didn't bring back in [this commit](https://github.com/dotnet/runtime/pull/85846/commits/5bdf838de2ccfffae79cbdd752b37030ab5ef2a0) from before I removed that compat infra.

**Comment by joperezr** [comment]
> This is now ready for review. The final decision on in/out API and justification can be found here: [MSFT internal xls](https://microsoft.sharepoint.com/:x:/r/teams/netfx/corefx/_layouts/15/Doc.aspx?sourcedoc=%7BC95246F3-E8C6-4911-9B27-B99C9BEEEE85%7D&file=extensions-compat-usage.xlsx&action=default&mobileredirect=true&share=IQHzRlLJxugRSZsnuZyb7u6FAQe-PcNae_gSYRixThXzBmY) / [public gist](https://gist.github.com/ericstj/c06988661190ec62736c748316692675) You can also see the differences with what we didn't bring back in [this commit](https://github.com/dotnet/runtime/pull/85846/commits/5bdf838de2ccfffae79cbdd752b37030ab5ef2a0) from before I removed that compat infra.

This is awesome, thanks a bunch for driving this @ericstj.

@geeknoid can we do a quick skim on that xls/gist to make sure all of the decisions there make sense?

@RussKie I know that you were previously doing a porting exercise of a sample that required some of these APIs to be brought back, and you used a private build at the time that Eric shared with you. Can we try that exercise again with the latest version of the packages from this PR to ensure everything still works as expected?

**Comment by ericstj** [comment]
You can find a private build of this PR from https://devdiv.visualstudio.com/DevDiv/_artifacts/feed/ericstj-share with the version `8.0.0-preview.6.extCompat.1`

**Comment by ericstj** [comment]
@tarekgh can you have a look at the final shape the Source changes are in and let me know if you’d like anything to change?

File: src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleLoggerExtensions.Obsolete.cs
Comment (tarekgh) [inline src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleLoggerExtensions.Obsolete.cs:116]: >NullChangeToken [](http://example.com/codeflow?start=29&length=15)

do we need to mark this as obsolete too? I am seeign we did that with `ConsoleLoggerSettingsAdapter`

File: src/libraries/Microsoft.Extensions.Logging.Debug/src/DebugLogger.cs
Comment (tarekgh) [inline src/libraries/Microsoft.Extensions.Logging.Debug/src/DebugLogger.cs:34]: >Everything [](http://example.com/codeflow?start=15&length=10)

nit: I agree the old comment is unclear but it was providing extra info that LogLevel.Debug is the highest level before `None` and that what refer to filter in the old comment.

File: src/libraries/Microsoft.Extensions.Options/src/OptionsWrapper.cs
Comment (tarekgh) [inline src/libraries/Microsoft.Extensions.Options/src/OptionsWrapper.cs:6]: why adding these here? wasn't the code compiling before adding them?

File: src/libraries/Microsoft.Extensions.Primitives/src/InplaceStringBuilder.cs
Comment (tarekgh) [inline src/libraries/Microsoft.Extensions.Primitives/src/InplaceStringBuilder.cs:11]: >[DebuggerDisplay("Value = {_value}")] [](http://example.com/codeflow?start=4&length=37)

it is interesting to keep the debug display while we are obsoleting the struct. I am fine either way though.

File: src/libraries/Microsoft.Extensions.Primitives/src/InplaceStringBuilder.cs
Comment (tarekgh) [inline src/libraries/Microsoft.Extensions.Primitives/src/InplaceStringBuilder.cs:41]: >_offset [](http://example.com/codeflow?start=20&length=7)

this is interesting as it is not thread safe :-)

**Review by tarekgh — APPROVED** [review]
LGTM. nice to see returning the compatibility back. 

Is there anything else left we decided not to bring?

File: src/libraries/Microsoft.Extensions.Options/src/OptionsWrapper.cs
Comment (ericstj) [reply src/libraries/Microsoft.Extensions.Options/src/OptionsWrapper.cs:6]: These were here from when I previously added obsolete API.  Missed cleaning these up.  Thanks!

File: src/libraries/Microsoft.Extensions.Primitives/src/InplaceStringBuilder.cs
Comment (ericstj) [reply src/libraries/Microsoft.Extensions.Primitives/src/InplaceStringBuilder.cs:41]: This type is not thread safe.

File: src/libraries/Microsoft.Extensions.Primitives/src/InplaceStringBuilder.cs
Comment (ericstj) [reply src/libraries/Microsoft.Extensions.Primitives/src/InplaceStringBuilder.cs:11]: I just pulled this file exactly as it was from the last shipping release.  https://github.com/dotnet/extensions/blob/release/3.1/src/Primitives/src/InplaceStringBuilder.cs

**Comment by ericstj** [comment]
> Is there anything else left we decided not to bring?

The XLS and gist linked above show what we didn't bring back.

File: src/libraries/Microsoft.Extensions.Logging.Debug/src/DebugLogger.cs
Comment (ericstj) [reply src/libraries/Microsoft.Extensions.Logging.Debug/src/DebugLogger.cs:34]: I don't think so.  This type used to have a member named filter.  That was removed but the comment was not updated: https://github.com/dotnet/extensions/blob/ab493da045ef1e645ff52675adb177ba7a503ce0/src/Logging/Logging.Debug/src/DebugLogger.cs#L46-L50

File: src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleLoggerExtensions.Obsolete.cs
Comment (ericstj) [reply src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleLoggerExtensions.Obsolete.cs:116]: No, it's private.  I marked `ConsoleLoggerSettingsAdapter` obsolete because it needed to use other obsolete API.  This type does not, so it doesn't need obsolete.

**Pull request merged**
